### PR TITLE
fix: Unexpected duplication of comments

### LINKED ISSUE #16106: [Bug]: duplicated comment after `path.remove()`
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Duplicated comments occurs after using `path.remove()`

### Input code

```js
import {parse} from '@babel/parser';
import traverse from '@babel/traverse';

const ast = parse(`
    // update-tsconfig-file
    const a = 5;
    // const updateTSConfig = require('../update-tsconfig')
    // https://nextjs.org/docs/app/building-your-application/upgrading/from-vite
    const getValue = (a) => a.value;
    getValue();
`);

traverse(ast, {
    VariableDeclarator(path) {
        if (path.node.id.name === 'a')
          path.remove();
    }
});
```

Comments duplicated, it should be 3 of them, not 4:

image

Any ways for such code:

```js
import {parse} from '@babel/parser';
import traverse from '@babel/traverse';

const ast = parse(`
    // update-tsconfig-file
    const a = 5;
    // const updateTSConfig = require('../update-tsconfig')
    // https://nextjs.org/docs/app/building-your-application/upgrading/from-vite
    const getValue = (a) => a.value;
    getValue();
`);

traverse(ast, {
    VariableDeclarator(path) {
        if (path.node.id.name === 'a')
          path.remove();
    },
    Program: {
        exit(path) {
            console.log(path + '');
        }
    }
});
```

Output will be correct:

image

So the problem that occurs in `@babel/traverse` is fixed by `@babel/generator`, but would be great if there was no duplication.

### Configuration file name

_No response_

### Configuration

_No response_

### Current and expected behavior

Current - 4 comments;
Expected - 3 comments, no duplicates;

### Environment

Latest Babel

### Possible solution

_No response_

### Additional context

It works in ast explorer, so as I understand, [`@babel/generator`](https://babeljs.io/docs/babel-generator) fixes all duplicates.

https://astexplorer.net/#/gist/7ed5bbcb21e2a3e96e609060936a41c2/3ba5a27a954d2a3e00aa5bcbaa359780ba5550e4

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #16106 
| Patch: Bug Fix?          | √
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | √
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

### CODE REVIEW

File: packages/babel-generator/src/printer.ts [hunk 1]
```diff
@@ -824,6 +824,10 @@ class Printer {
       if (i < len - 1) separator?.();
 
       if (opts.statement) {
+        if (!node.trailingComments?.length) {
```
Comment (liuxingbaoyu) [inline packages/babel-generator/src/printer.ts:827]: This fixes the extra new line.
```
{
  var t = 20; /*
               * This is trailing comment
               */
}

{
  var tt = 20; /*
                * This is trailing comment
                */
}
```

File: packages/babel-traverse/src/path/comments.ts [hunk 2]
```diff
@@ -48,11 +48,9 @@ export function shareCommentsWithSiblings(this: NodePath) {
...
   if (!toRemove) return list;
-  let lastFoundIndex = -1;
+  const set = new Set(toRemove);
   return list.filter(el => {
-    const i = toRemove.indexOf(el, lastFoundIndex);
-    if (i === -1) return true;
-    lastFoundIndex = i;
+    return !set.has(el);
   });
 }
```
Comment (liuxingbaoyu) [inline packages/babel-traverse/src/path/comments.ts:55]: The problem is that the order may be reversed.
[1,2] [2,1]

File: packages/babel-generator/test/index.js [hunk 3]
```diff
@@ -1492,6 +1492,26 @@ describe("programmatic generation", function () {
   val
 );`);
     });
+
+    it("multi-line leading comment after return 2", () => {
```
Comment (liuxingbaoyu) [inline packages/babel-generator/test/index.js:1496]: This test has nothing to do with this PR, it just ensures that https://github.com/babel/babel/pull/16104 does not cause regressions.

**Review by JLHwung — APPROVED** [review]

File: babel.config.js [hunk 4]
```diff
@@ -304,6 +305,16 @@ module.exports = function (api) {
     ].filter(Boolean),
   };
 
+  if (jestSnapshot) {
```
Comment (nicolo-ribaudo) [inline babel.config.js:308]: Can we configure Jest to just not load the Babel config for snapshots?

**Review by nicolo-ribaudo — APPROVED** [review]

File: babel.config.js (hunk 4 above)
Comment (liuxingbaoyu) [reply babel.config.js:308]: https://github.com/jestjs/jest/blob/2c3d2409879952157433de215ae0eee5188a4384/packages/jest-snapshot/src/utils.ts#L363-L368

This doesn't seem to be possible at the moment.

Comment (liuxingbaoyu) [reply babel.config.js:308]: To be honest, I even want to open a pr in jest to avoid using `@babel/core`.

**Pull request merged**